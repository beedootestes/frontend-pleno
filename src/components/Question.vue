<template>
  <div class="question">
    <div class="header-question">
      <div class="icon-help"><HelpIcon :size="40" /></div>
      <p class="title">{{ question.text }}</p>
      <p class="beetcoins">({{question.beetcoins}}) moedas</p>
    </div>

    <div class="alternatives">
      <AlternativeQuestion
        class="item"
        @click="selectAlternative(question.id, alternative.id)"
        v-for="alternative in question.alternatives"
        :key="alternative.id"
        :alternative="alternative"
        :correctAnswerId="correctAnswerId"
        :selected="alternative.id === selectedAlternative" />
    </div>
  </div>
</template>

<script setup lang="ts">
  import HelpIcon from 'vue-material-design-icons/Help.vue';
  import AlternativeQuestion from "./AlternativeQuestion.vue";
  import { PropType, ref } from "vue";
  import type { Question } from "../types/question.ts";

  const selectedAlternative = ref<Number>(0)

  const emits = defineEmits(['update:modelValue'])

  const selectAlternative = (questionId, alternativeId) => {
    selectedAlternative.value = alternativeId
    emits('update:modelValue', {questionId, alternativeId})
  }

  defineProps({
    question: {
      type: Object as PropType<Question>,
      required: true
    },

    correctAnswerId: {
      type: Number,
      required: true,
      default: 0
    }
  })
</script>

<style scoped lang="scss">
  .question {
    .header-question {
      display: flex;
      align-items: center;
      box-sizing: border-box;
      padding: 20px 0;

      .icon-help {
        background-color: $bg-answer-detail;
        box-sizing: border-box;
        padding: 15px;
        border-radius: 50%;
        color: #fff;
        width: 73px;
        margin-right: 15px;
      }

      p {
        margin: 0;
        margin-block-end: 0;
        margin-block-start: 0;

        &:nth-of-type(1) { margin-right: 5px; }
        &:nth-of-type(2) { font-size: 12px; }
      }
    }

    .alternatives {
      .item { margin-bottom: 5px; }

      &:last-child { margin-bottom: 0; }
    }
  }
</style>