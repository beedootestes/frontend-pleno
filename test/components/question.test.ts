import {describe, expect, test, vi} from "vitest";
import Question from "../../src/components/Question.vue";
import {mount, VueWrapper} from "@vue/test-utils";
import {PropType} from "vue";


describe('Test Question.vue', () => {
  test('mount component', async () => {
    expect(Question).toBeTruthy()

    const wrapper: VueWrapper<Question> = mount(Question, {
      props: {
        question: {
          "beetcoins": 50,
          "id": 1,
          "text": "Question text",
          "parent_id": 15,
          "alternatives": [{
            "id": 11,
            "text": "Alternative one.",
            "question_id": 1,
            "is_correct": true
          }, {
            "id": 12,
            "text": "alternative two.",
            "question_id": 1,
            "is_correct": false
          }]
        },
        correctAnswerId: 0
      }
    })

    expect(wrapper.html()).toMatchSnapshot()

    const title = wrapper.find('.title')
    const beetcoins = wrapper.find('.beetcoins')

    expect(title.text()).toStrictEqual('Question text')
    expect(beetcoins.text()).toStrictEqual('(50) moedas')
  })

  test('test if when click in a alternative call a function selectAlternative', async () => {
    const wrapper: VueWrapper<Question> = mount(Question, {
      props: {
        question: {
          "beetcoins": 50,
          "id": 1,
          "text": "Question text",
          "parent_id": 15,
          "alternatives": [{
            "id": 11,
            "text": "Alternative one.",
            "question_id": 1,
            "is_correct": true
          }, {
            "id": 12,
            "text": "alternative two.",
            "question_id": 1,
            "is_correct": false
          }]
        },
        correctAnswerId: 0
      }
    })

    const spy = vi.spyOn(wrapper.vm, 'selectAlternative')
    const firstAlternative = wrapper.find('.item')
    await firstAlternative.trigger('click')
    expect(spy).toHaveBeenCalledTimes(1)
  })


  test('test if when click in a alternative set alternative selected', async () => {
    const wrapper: VueWrapper<Question> = mount(Question, {
      props: {
        question: {
          "beetcoins": 50,
          "id": 1,
          "text": "Question text",
          "parent_id": 15,
          "alternatives": [{
            "id": 11,
            "text": "Alternative one.",
            "question_id": 1,
            "is_correct": true
          }, {
            "id": 12,
            "text": "alternative two.",
            "question_id": 1,
            "is_correct": false
          }]
        },
        correctAnswerId: 0
      }
    })

    expect(wrapper.vm.selectedAlternative).toStrictEqual(0)
    const firstAlternative = wrapper.find('.item')
    await firstAlternative.trigger('click')
    expect(wrapper.vm.selectedAlternative).toStrictEqual(11)
  })


  test('test if when click in a alternative and emit update:modelValue event', async () => {
    const wrapper: VueWrapper<Question> = mount(Question, {
      props: {
        question: {
          "beetcoins": 50,
          "id": 1,
          "text": "Question text",
          "parent_id": 15,
          "alternatives": [{
            "id": 11,
            "text": "Alternative one.",
            "question_id": 1,
            "is_correct": true
          }, {
            "id": 12,
            "text": "alternative two.",
            "question_id": 1,
            "is_correct": false
          }]
        },
        correctAnswerId: 0
      }
    })

    expect(wrapper.vm.selectedAlternative).toStrictEqual(0)
    const firstAlternative = wrapper.find('.item')
    await firstAlternative.trigger('click')
    expect(wrapper.emitted('update:modelValue'))
    expect(wrapper.emitted('update:modelValue')).toStrictEqual([[{questionId: 1, alternativeId: 11}]])
  })


  test('test if when click in a alternative and set a correctAnswerId prop', async () => {
    const wrapper: VueWrapper<Question> = mount(Question, {
      props: {
        question: {
          "beetcoins": 50,
          "id": 1,
          "text": "Question text",
          "parent_id": 15,
          "alternatives": [{
            "id": 11,
            "text": "Alternative one.",
            "question_id": 1,
            "is_correct": true
          }, {
            "id": 12,
            "text": "alternative two.",
            "question_id": 1,
            "is_correct": false
          }]
        },
        correctAnswerId: 0
      }
    })

    expect(wrapper.vm.selectedAlternative).toStrictEqual(0)
    await wrapper.setProps({ correctAnswerId: 11 })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
