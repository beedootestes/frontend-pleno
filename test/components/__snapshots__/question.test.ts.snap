// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Question.vue > mount component 1`] = `
"<div data-v-5ab2d74e=\\"\\" class=\\"question\\">
  <div data-v-5ab2d74e=\\"\\" class=\\"header-question\\">
    <div data-v-5ab2d74e=\\"\\" class=\\"icon-help\\"><span data-v-5ab2d74e=\\"\\" aria-hidden=\\"true\\" class=\\"material-design-icon help-icon\\" role=\\"img\\"><svg fill=\\"currentColor\\" class=\\"material-design-icon__svg\\" width=\\"40\\" height=\\"40\\" viewBox=\\"0 0 24 24\\"><path d=\\"M10,19H13V22H10V19M12,2C17.35,2.22 19.68,7.62 16.5,11.67C15.67,12.67 14.33,13.33 13.67,14.17C13,15 13,16 13,17H10C10,15.33 10,13.92 10.67,12.92C11.33,11.92 12.67,11.33 13.5,10.67C15.92,8.43 15.32,5.26 12,5A3,3 0 0,0 9,8H6A6,6 0 0,1 12,2Z\\"><!--v-if--></path></svg></span></div>
    <p data-v-5ab2d74e=\\"\\" class=\\"title\\">Question text</p>
    <p data-v-5ab2d74e=\\"\\" class=\\"beetcoins\\">(50) moedas</p>
  </div>
  <div data-v-5ab2d74e=\\"\\" class=\\"alternatives\\">
    <div data-v-58d5336b=\\"\\" data-v-5ab2d74e=\\"\\" class=\\"answer question-item item\\">
      <div data-v-58d5336b=\\"\\">Alternative one.</div>
      <div data-v-58d5336b=\\"\\">
        <!--v-if-->
      </div>
    </div>
    <div data-v-58d5336b=\\"\\" data-v-5ab2d74e=\\"\\" class=\\"answer question-item item\\">
      <div data-v-58d5336b=\\"\\">alternative two.</div>
      <div data-v-58d5336b=\\"\\">
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;

exports[`Test Question.vue > test if when click in a alternative and set a correctAnswerId prop 1`] = `
"<div data-v-5ab2d74e=\\"\\" class=\\"question\\">
  <div data-v-5ab2d74e=\\"\\" class=\\"header-question\\">
    <div data-v-5ab2d74e=\\"\\" class=\\"icon-help\\"><span data-v-5ab2d74e=\\"\\" aria-hidden=\\"true\\" class=\\"material-design-icon help-icon\\" role=\\"img\\"><svg fill=\\"currentColor\\" class=\\"material-design-icon__svg\\" width=\\"40\\" height=\\"40\\" viewBox=\\"0 0 24 24\\"><path d=\\"M10,19H13V22H10V19M12,2C17.35,2.22 19.68,7.62 16.5,11.67C15.67,12.67 14.33,13.33 13.67,14.17C13,15 13,16 13,17H10C10,15.33 10,13.92 10.67,12.92C11.33,11.92 12.67,11.33 13.5,10.67C15.92,8.43 15.32,5.26 12,5A3,3 0 0,0 9,8H6A6,6 0 0,1 12,2Z\\"><!--v-if--></path></svg></span></div>
    <p data-v-5ab2d74e=\\"\\" class=\\"title\\">Question text</p>
    <p data-v-5ab2d74e=\\"\\" class=\\"beetcoins\\">(50) moedas</p>
  </div>
  <div data-v-5ab2d74e=\\"\\" class=\\"alternatives\\">
    <div data-v-58d5336b=\\"\\" data-v-5ab2d74e=\\"\\" class=\\"correct question-item item\\">
      <div data-v-58d5336b=\\"\\">Alternative one.</div>
      <div data-v-58d5336b=\\"\\"><span data-v-58d5336b=\\"\\" class=\\"correct-icon material-design-icon check-circle-icon correct-icon\\" aria-hidden=\\"true\\" role=\\"img\\"><svg fill=\\"#228B21\\" class=\\"material-design-icon__svg\\" width=\\"45\\" height=\\"45\\" viewBox=\\"0 0 24 24\\"><path d=\\"M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z\\"><!--v-if--></path></svg></span></div>
    </div>
    <div data-v-58d5336b=\\"\\" data-v-5ab2d74e=\\"\\" class=\\"incorrect question-item item\\">
      <div data-v-58d5336b=\\"\\">alternative two.</div>
      <div data-v-58d5336b=\\"\\"><span data-v-58d5336b=\\"\\" aria-hidden=\\"true\\" class=\\"material-design-icon alpha-x-circle-icon\\" role=\\"img\\"><svg fill=\\"#FF0000\\" class=\\"material-design-icon__svg\\" width=\\"45\\" height=\\"45\\" viewBox=\\"0 0 24 24\\"><path d=\\"M9,7L11,12L9,17H11L12,14.5L13,17H15L13,12L15,7H13L12,9.5L11,7H9M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2Z\\"><!--v-if--></path></svg></span></div>
    </div>
  </div>
</div>"
`;
